azure-kusto-data/azure/kusto/data/client.py

```diff
@@ -1,962 +1,972 @@
+                request_headers["x-ms-client-request-id"] = properties.client_request_id
+            if properties.application is not None:
+                request_headers["x-ms-app"] = properties.application
+            if properties.user is not None:
+                request_headers["x-ms-user"] = properties.user
+            if properties.get_option(ClientRequestProperties.no_request_timeout_option_name, False):
+                timeout = KustoClient._mgmt_default_timeout
+            else:
+                timeout = properties.get_option(ClientRequestProperties.request_timeout_option_name, timeout)
+
+        timeout = (timeout or KustoClient._mgmt_default_timeout) + KustoClient._client_server_delta
+
+        self.json_payload = json_payload
+        self.request_headers = request_headers
+        self.timeout = timeout
+
+
+class HTTPAdapterWithSocketOptions(requests.adapters.HTTPAdapter):
+    def __init__(self, *args, **kwargs):
+        self.socket_options = kwargs.pop("socket_options", None)
+        self.pool_maxsize = kwargs.pop("pool_maxsize", None)
+        self.max_retries = kwargs.pop("max_retries", None)
+        super(HTTPAdapterWithSocketOptions, self).__init__(*args, **kwargs)
+
+    def init_poolmanager(self, *args, **kwargs):
+        if self.socket_options is not None:
+            kwargs["socket_options"] = self.socket_options
+        super(HTTPAdapterWithSocketOptions, self).init_poolmanager(*args, **kwargs)
+
+
+class _KustoClientBase(abc.ABC):
+    API_VERSION = "2019-02-13"
+
+    _mgmt_default_timeout = timedelta(hours=1, seconds=30)
+    _query_default_timeout = timedelta(minutes=4, seconds=30)
+    _streaming_ingest_default_timeout = timedelta(minutes=10)
+
+    _aad_helper: _AadHelper
+
+    def __init__(self, kcsb: Union[KustoConnectionStringBuilder, str], is_async):
+        self._kcsb = kcsb
+        self._proxy_url: Optional[str] = None
+        if not isinstance(kcsb, KustoConnectionStringBuilder):
+            self._kcsb = KustoConnectionStringBuilder(kcsb)
+        self._kusto_cluster = self._kcsb.data_source
+
+        # notice that in this context, federated actually just stands for add auth, not aad federated auth (legacy code)
+        self._aad_helper = _AadHelper(self._kcsb, is_async) if self._kcsb.aad_federated_security else None
+
+        # Create a session object for connection pooling
+        self._mgmt_endpoint = "{0}/v1/rest/mgmt".format(self._kusto_cluster)
+        self._query_endpoint = "{0}/v2/rest/query".format(self._kusto_cluster)
+        self._streaming_ingest_endpoint = "{0}/v1/rest/ingest/".format(self._kusto_cluster)
+        self._request_headers = {
+            "Accept": "application/json",
+            "Accept-Encoding": "gzip,deflate",
+            "x-ms-client-version": "Kusto.Python.Client:" + VERSION,
+            "x-ms-version": self.API_VERSION,
+        }
+
+    def set_proxy(self, proxy_url: str):
+        self._proxy_url = proxy_url
```

Comment:
If we use the `aad_helper` to maintain the `proxy_dict`, why not use it for the `proxy_url`?

I think it would be cleaner to remove the `_proxy_url` from the client, and instead the `aad_helper` would maintain the `proxy_url`. I'd then remove the `proxy_dict` from being a property of `aad_helper`, and instead a method that dynamically returns the `proxy_dict` based on its `proxy_url` property.

Reply:
Sorry, my mistake - it's not in aad_helper but in the token provider.